nxpprog.py

```diff
@@ -404,73 +404,75 @@
                 'OK': b'OK\r\n'}
 
 
-def port_write_and_verify(port, payload, error_message="", debug_message=""):
-    bytes_sent = port.write(bytearray(payload))
-    if bytes_sent != len(payload):
-        log(error_message)
-        return False
-    else:
-        log(debug_message)
-        return True
-
-
-def port_read(port, number_of_bytes):
-    return bytearray(port.read(number_of_bytes))
-
-
-def attempt_sync(port):
-    log(f'Attempting to sync with device at {port.port}')
-    time.sleep(.100)
-    port.dtr = True
-    port.rts = True
-    time.sleep(.100)
-    port.dtr = False
-    time.sleep(.100)
-    port.rts = False
-    time.sleep(1)
-    port_write_and_verify(port, LPC40XX_CHAR['Question'])
-    time.sleep(.100)
-    response = port_read(port, 14)
-    time.sleep(.100)
-    port_write_and_verify(port, LPC40XX_CHAR['Synchronized'])
-    time.sleep(.100)
-    response = port_read(port, 17)
-    port_write_and_verify(port, b'10000\r\n')
-    time.sleep(.100)
-    response = port_read(port, 4)
-    return response == LPC40XX_CHAR['Synchronized']
-
-
-def find_lpc_port(device=None):
-    log('Looking for ports with LPC40xx devices')
-    port = None
-    if device:
-        serial_device_list = [
-            {
-                "device": device
-            }
-        ]
-    else:
-        serial_device_list = list(port_list.comports())
+class AutoLPCPortFinder:
+    def __init__(self, device=None):
+        self.device = device
+
+    def port_write_and_verify(self, port, payload, error_message="", debug_message=""):
+        bytes_sent = port.write(bytearray(payload))
+        if bytes_sent != len(payload):
+            log(error_message)
+            raise SerialException(f"Unsuccessful writing {payload} to {port}")
+        else:
+            log(debug_message)
+
+
+    def port_read(self, port, number_of_bytes):
+        return bytearray(port.read(number_of_bytes))
+
+    def find_lpc_port(self):
+        log('Looking for ports with LPC40xx devices')
+        port = None
+        if self.device:
+            serial_device_list = [
+                {
+                    "device": self.device
+                }
+            ]
+        else:
+            serial_device_list = list(port_list.comports())
+
+        for port_info in serial_device_list:
+            try:
+                port = serial.Serial(
+                    port=port_info.device,
+                    baudrate=115200,
+                    parity=serial.PARITY_NONE,
+                    stopbits=serial.STOPBITS_ONE,
+                    bytesize=serial.EIGHTBITS,
+                    timeout=3)
+                sync_successful = self.attempt_sync(port)
+                if sync_successful:
+                    port.close()
+                    return port_info.device
+            except serial.SerialException:
+                log(f'excpetion on port {port}')
```

Comment:
If you checkout my Hyperload example of this you will notice that this is printed out using `log.debug` so that this message only gets printed out when testing the script with the `--verbose` flag. If you have a bunch of serial ports that aren't connect to anything, this exception will make the user think that something is wrong with their computer when in reality, this is expected behavior. This is especially annoying on windows machines using WSL where windows has a `/dev/ttySx` for each possible COM port (which there are up to 100 of them). If there isn't a corresponding COM port it returns with a serial exception. This is annoying because, if your port comes up as 85, you will see 84 "exception" messages before programming cant start.

So we should look into using a log statement that is masked when not on verbose mode, but that may be better for another commit. So you if we have a way of debug logging it in this file, use it, or simply remove this line and add an "issue" to bring it back.

Reply:
I will remove the exception, and make an issue to have all logging use the `logging` python module, I can then specify the debug option like you did in Hyperload like below

```python
if args.verbose:
  logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
else:
  logging.basicConfig(stream=sys.stdout, level=logging.INFO)
```

Reply:
That would be good. Please do that.

Reply:
This change was added to #14